include/constant_medium.hpp

```diff
@@ -62,7 +60,8 @@ class constant_medium {
     /// Distance between the two hitpoints affect of probability
     /// of the ray hitting a smoke particle
     const auto distance_inside_boundary = (rec2.t - rec1.t) * ray_length;
-    const auto hit_distance = neg_inv_density * sycl::log(rng.float_t());
+    auto rng = LocalPseudoRNG { toseed(r.direction()) };
```

Comment:
Ah, I realize you have removed the context passed everywhere and replaced basically the random generator with some local computation.
What was the rationale?

Reply:
As getting a value from the RNG update its internal state, having a shared RNG creates a very long chain of read after write dependency that prevent the HLS compiler to parallelize the otherwise independent computation steps.

Reply:
For example https://xilinx.github.io/Vitis_Accel_Examples/2020.2/html/dependence_inter.html

Reply:
Otherwise, perhaps something like this pseudo HLS-SYCL code:
```c++
auto rng(auto local_seed = std::source_location::current()) {
  return [=] {
     static auto state = local_seed.line();
     #pragma HLS dependence variable=state inter false
     state = crunch(state);
     return mix(state);
  };
}
```
or
```c++
auto rng(auto local_seed = std::source_location::current()) {
  return [state = local_seed.line()] mutable {
     #pragma HLS dependence variable=state inter false
     state = crunch(state);
     return mix(state);
  };
}
```
Perhaps we need extension https://github.com/intel/llvm/pull/3746 to have the big picture working.

Reply:
@aisoard, @yu810226 any feedback?

Reply:
Above is basically a good idea, if I understand the intention, to separate the state into multiple independent RNGs. I'm not sure what you're trying to achieve with the false dependence pragma, however.